### ISSUE TITLE
match_wild_err_arm incorrectly recommends for constant variables.

### ISSUE THREAD

**Issue by Mythra** [rust-lang/rust-clippy#10635]
### Summary

When a constant is set to an inline match statement which panics on error (we can't use `.expect()`, and `.unwrap()` in const yet, so it has to be a match to panic), clippy incorrectly fires a `match_wild_err_arm` statement to code that would not work.

### Lint Name

match_wild_err_arm

### Reproducer

Take the following rust code (which depends on the `time` crate):

```rust
use time;
// $X has all timestamps start from january 1st, 2000, not normal epoch of 1970.
const COOLER_EPOCH: time::OffsetDateTime = match time::OffsetDateTime::from_unix_timestamp(946_684_800) {
	Ok(val) => val,
	Err(_) => panic!("`COOLER_EPOCH` is not valid :( you darn screwed up partner"),
};
```

clippy generates the following error:

```
error: `Err(_)` matches all errors
 --> src/main.rs:9:2
  |
9 |     Err(_) => panic!("`COOLER_EPOCH` is not valid :( you darn screwed up partner"),
  |     ^^^^^^
  |
  = note: match each error separately or use the error output, or use `.expect(msg)` if the error case is unreachable
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_wild_err_arm
  = note: `-D clippy::match-wild-err-arm` implied by `-D warnings`
```

This code however is in a const context, and I cannot replace it with an expect, or it will end up generating an error:

```
error[[E0015]](https://doc.rust-lang.org/stable/error_codes/E0015.html): cannot call non-const fn `Result::<OffsetDateTime, time::error::ComponentRange>::expect` in constants
  --> src/main.rs:9:101
   |
9  | const COOLER_EPOCH: time::OffsetDateTime = time::OffsetDateTime::from_unix_timestamp(946_684_800).expect("blah");
   |                                                                                                     ^^^^^^^^^^^^^^
   |
   = note: calls in constants are limited to constant functions, tuple structs and tuple variants
```

### Version

```text
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-unknown-linux-gnu
release: 1.68.2
LLVM version: 15.0.6
```
